[discussion] coordinate frames for industrial manipulators

gavanderhoorn opened the pull request:
I'll never merge this, but PRs are more suited for discussions than single commits.

gavanderhoorn committed:
Current state.

gavanderhoorn committed:
Update comment on 'toolN' naming of tool frames.

gavanderhoorn committed:
Reword expected frame authority section.

gavanderhoorn committed:
Be consistent in calling everything a manipulator.

gavanderhoorn committed:
Phrasing.

gavanderhoorn committed:
Provide example of when RST might not be the frame authority.

gavanderhoorn commented on rep-xxxx.rst:
```diff
@@ -0,0 +1,248 @@
+=========
+
+#. are there robots that do not have ``tool0`` defined as a child of ``flange`` (I seem to remember some brand having ``tool0`` defined relative to the last link/joint, which was not the flange)? If not: make ``tool0`` child of ``flange``. Otherwise keep ``tool0`` a sibling of ``flange``.
+#. should we discourage naming additional frames ``toolN`` (ie: ``tool1``, ``tool2``, etc)? Such names carry no semantics (frame purpose only known out-of-band), and all names in ROS should be as descriptive as possible.
```
@shaun-edwards comment (on ec18f94e2a008a44840d680a8693105b92a3500a):
> Some robots store tool poses in an array.  Tool0 actually refers to the too at index zero, typically defined as all zeros.  Tool1 implies the tool at index 1.

gavanderhoorn replied:
There are indeed many mfgs that handle their (tool)frames like that. Are you suggesting we don't discourage `toolN` naming?

I feel our naming should not follow / be (too much) influenced by what industrial controllers do. `tool0` is already unfortunate, as I've seen many people starting to introduce `tool1`, `tool2`, etc. Naming in ROS should be done with semantics in mind, and those names do not have any semantics.

Of course consumers of our models are essentially free to do whatever they want in their own derived models, but I'd really like to recommend against using `toolN` frame names (it's almost like naming a variable `a` and `b`).

gavanderhoorn replied:
<del>Note btw that the question is actually wrong (copied old version of comment)</del>: in the new setup (with `flange` and `tool0` as separate frames), `tool0` does not get any children. Only `flange` has child frames. So any additional tool frames should be made *siblings* of `tool0` (and they don't even have to have `flange` or 'the last link' as their parent necessarily).

---

In the meantime I've updated the text, but I've left this comment about `toolN`s being siblings.

JeremyZoss replied:
In general, I think `tool1`...`toolN` are application tool-frames, and shouldn't be included in the base robot URDF.  But we can note that these names should be reserved for that purpose (representing equivalent robot-controller frames).

My view is that `tool1`...`toolN` are children of `tool0`.  `tool0` typically represents the "all-zeros" tool frame.  By definition, then, any user-defined tool frames are defined relative to the nominal ("all-zeros"/`tool0`) frame.

In reality, these frames are typically used to represent different end-effectors that are attached to the robot flange.  They may be more appropriately represented in ROS as eef-specific TCP frames, not as `tool1`, etc.

gavanderhoorn replied:
@JeremyZoss wrote:
> In general, I think `tool1`...`toolN` are application tool-frames, and shouldn't be included in the base robot URDF.  But we can note that these names should be reserved for that purpose (representing equivalent robot-controller frames).
> 
> [..]
> 
> In reality, these frames are typically used to represent different end-effectors that are attached to the robot flange.  They may be more appropriately represented in ROS as eef-specific TCP frames, not as `tool1`, etc.

I agree with your last point: I don't think using (what are essentially) indices for names is a good idea. ROS allows you to use names for these kind of things, and we should use/encourage that.

The question was thus: do people feel strongly about allowing `toolN` names or not? Personally I would like to strongly discourage it, but as I also noted, consumers of our urdfs/xacros can do whatever they want of course.

> My view is that `tool1`...`toolN` are children of `tool0`.  `tool0` typically represents the "all-zeros" tool frame.  By definition, then, any user-defined tool frames are defined relative to the nominal ("all-zeros"/`tool0`) frame.

I'm not sure I share this view. `tool0` just happens to be a tool frame with all entries "at 0". But I could just as easily define another frame that is configured as such, and that wouldn't make it the parent of all other tool frames.

Tool frames are typically stated in manuals as being relative to the "flange". An all-zeros frame happens to lie on the flange for most robots (but this is not guaranteed, IIRC, see the first question).

I believe @shaun-edwards chose the `tool0` name as it has special meaning on ABB controllers (ie: the default one). Fanucs fi don't have that.

Finally: making everything 'after' `tool0` a child of it has its problems (due to it not being REP-103 compliant), which is why `flange` was introduced.

gavanderhoorn commented:
@JeremyZoss, @Levi-Armstrong, @Jmeyer1292: if you have the time, I would appreciate your feedback on this. It's not finished yet, but early feedback is good.

gavanderhoorn committed:
Not using these conventions is ok, but discouraged.

gavanderhoorn committed:
Add example hierarchy with EEF.

gavanderhoorn commented on rep-xxxx.rst:
```diff
@@ -0,0 +1,268 @@
+          │ └ eef_base_link
+          │   └ ..
+          │     └ eef_link_N
+          └ eef_tcp
```
I'm not sure here whether any `tcp` frame should be a child of `flange` or not.

Many methods of 'teaching' tool frames will (in the end) be relative to the flange of the robot, so making all tool frames (including `tool0`) children of `flange` could make sense.

However that would also imply that the origin of `flange` must be exactly what it is on the real robot, which may not always be known or documented, and makes the frame not only a convenience (ie: an attachment point for EEF xacro models), but promotes it to an entity that (should accurately) model(s) a physical part of the real robot.

gavanderhoorn replied:
(slightly out-of-scope here) I didn't make `eef_tcp` a child of `eef_base_link` as that is not how I believe most (all?) ind robots do this. Tool frames will always be relative to the flange, and by doing it this way, we also keep the transform `base -> eef_tcp` relatively uncomplicated.

JeremyZoss replied:
I think `link_N` should be mostly internal, and that all external frames should reference either `flange` or `tool0`.  Under that logic, `eef_tcp` is wrongly placed here.  Some more discussion is required about how TCP is handled, though.

  - There is some advantage to deriving `eef_tcp` from `eef_base_link`.  This allows more modularity in packaging of eef URDFs, so that a new "tool" can be attached with a single joint, defining both geometry and TCP.

  - Should TCP be attached to `flange` or `tool0`?  This may depend on how closely your ROS application is coupled to the controller.  It might be handy to do planning relative to `base` and `tool0` if passing Cartesian coords to robot, but ROS-centric planning might prefer to use `base_link` and `flange`.
  - Can we at least provide recommendations on how to assign TCP?  As a formal section, not just a footnote below this diagram.

gavanderhoorn replied:
@JeremyZoss wrote:
> I think `link_N` should be mostly internal, and that all external frames should reference either `flange` or `tool0`.  Under that logic, `eef_tcp` is wrongly placed here.  Some more discussion is required about how TCP is handled, though.
> 
>   - There is some advantage to deriving `eef_tcp` from `eef_base_link`.  This allows more modularity in packaging of eef URDFs, so that a new "tool" can be attached with a single joint, defining both geometry and TCP.

yes, but it would/could also complicate setting up xacros that can be calibrated. If we include tool frames in EEF models, the files will have to be edited (difficult if the EEF was distributed as a binary package (ie: non world-writable)) if the toolframe needs to be adjusted for some reason. If we keep tool frames out of the EEF xacro, users that want to reuse them can easily do calibration of the composition `robot+eef` and then add their new tool frame by attaching it to `robot_flange` in the composing xacro.

We could of course standardise naming of frames in EEF xacros, and provide something similar to `flange` (ie: an attachment point) in those xacros, but I expect more variability in EEFs than in robots, so that could be a bit much (for now at least).

gavanderhoorn replied:
@JeremyZoss wrote:
>  - Should TCP be attached to `flange` or `tool0`?  This may depend on how closely your ROS application is coupled to the controller.  It might be handy to do planning relative to `base` and `tool0` if passing Cartesian coords to robot, but ROS-centric planning might prefer to use `base_link` and `flange`.

I think it should be `flange` in any case, as `tool0` does not follow REP 103 (they typically have z+ forward) which complicates attaching additional links in a xacro/urdf (you need to introduce additional rotations to undo the x+->z+ change in `tool0`).

As to planning: I feel either:

 - planning libraries should be able to transform incoming poses to whatever the parent is of the configured EEF (MoveIt is a bit inconsistent with it, but it can do it, that's why EEF groups are configured in the MSA)
 - users should transform poses back to `tool0` (or whatever we decide is going to be the 'end link' (but see my reply to your later comment))

gavanderhoorn replied:
@JeremyZoss wrote:
>   - Can we at least provide recommendations on how to assign TCP?  As a formal section, not just a footnote below this diagram.

Can you clarify what you mean? I'm not sure I understand the comment.

JeremyZoss replied:
> > Can we at least provide recommendations on how to assign TCP? As a formal section, not just a footnote below this diagram.

> Can you clarify what you mean? I'm not sure I understand the comment.

I feel like most "real" applications (as opposed to simplified exercises) will try to control a robot's TCP.  It's not obvious/trivial what is the "typical" approach for creating such a TCP frame: naming convention, parent frame, etc, as evidenced by some of the discussions on this page.  My comment here means that it might be worth adding a section to discuss expectations related to the TCP frame.  It's not really mentioned in this writeup, except as listed in the "Single Arm Manipulator with EEF" diagram.  The diagram should (ideally) illustrate concepts discussed elsewhere in the writeup.

gavanderhoorn replied:
I'm not against adding something about TCP frames, but I'm not sure it's in the scope of this document per se. As with REP-105 (frames for mobile bases) and REP-120 (frames for humanoids) the intent is to define a nr of standardised frames and their semantics - but for industrial robots.

I think there is definitely enough to say about it, but perhaps in a separate REP?

gavanderhoorn commented on rep-xxxx.rst:
```diff
@@ -0,0 +1,268 @@
+Questions
+=========
+
+#. are there robots that do not have ``tool0`` defined as a child of ``flange`` (I seem to remember some brand having ``tool0`` defined relative to the last link/joint, which was not the flange)? If not: make ``tool0`` child of ``flange``. Otherwise keep ``tool0`` a sibling of ``flange``.
```
same question as for the *Single manipulator with EEF* here.

JeremyZoss replied:
I recommend we derive `tool0` from `flange`, not `link_6`.  Arguments can be made both ways, but there are many cases when the same offset may be required from `link_6` in both cases.  Less duplication/error if that offset is only defined once.  This can simply be a recommendation, not a requirement, so that unusual cases could derive `tool0` directly from the last link.

gavanderhoorn replied:
Ok, makes sense.

gavanderhoorn commented:
I'm not sure whether the discussion around how to model EEFs and their TCPs is out-of-scope here, but I would appreciate any comments on that as well, as it at least partially defines the semantics of `flange` and `tool0`.

JeremyZoss commented:
There is potential confusion with references to `base_link` and "kinematic chain"
  - You should clarify that we mean "___ROS___ kinematic chain" (no tie to robot kinematics)
  - Remove the phrase "and forms the root of the kinematic chain..." from the initial `base_link` paragraph.  The key here is that this link is based on the physical robot geometry.
  - Move the sentence "No special treatment ... root of kinematic chain..." from the `base_link` rationale section to a standalone paragraph above.  This is a recommendation for what to do with `base_link`, not a rationale for why we're doing that.

JeremyZoss commented:
What does the phrase "shall not be changed" mean in the `flange` and `tool0` sections?  For the most part, all frames in the stock robot-model URDFs are considered to be static and un-changing.

JeremyZoss commented:
We don't currently have a `flange` model in most ROS-I URDFs, and in some cases may be using `tool0` for similar purposes (e.g. [abb\irb2400](https://github.com/ros-industrial/abb/blob/kinetic-devel/abb_irb2400_support/urdf/irb2400_12_155_macro.xacro#L150)).  How do we intend to handle this transition?  Some concerns:
  - changes to tool0 may affect existing models
  - how do we flag robots that have "compliant" and "non-compliant" tool0 definitions?

JeremyZoss commented:
We should include a recommendation on where the kinematic chain should __end__.  You recommend it use `base_link` as the root (okay), but do we recommend `flange` or `tool0` as the tip?  I recommend `flange`, as it seems the equivalent of `base_link`.  `base_link` and `flange` are both ROS-centric, geometry-based frames.  `base` and `tool0` seem more like virtual frames for association with robot controller kinematics/positioning.

JeremyZoss commented:
Do you need to add a discussion about intermediate frame/link names?  In general, there is more freedom/variation here, and most external ROS code/URDF should reference the base/tool frames instead.  However, it may still be good to standardize where possible.  We could recommend use of `link_1`...`link_6`, with exceptions to follow manufacturer-specific joint naming (e.g. motoman).

JeremyZoss commented:
Do we need to recommend a standard naming convention for the TCP frame?  It's somewhat outside the scope of this document, since it won't be captured in the core robot URDF, but it's related to concerns about modularity and standardization.

JeremyZoss commented:
In your Dual Manipulator chart, I disagree with the placement of `base_link`.  It feels like a true dual-arm manipulator (e.g. SDA20) might have a `base_link` as shown, along with other potential joints (e.g. `torso`).  However, a welding cell with two standalone manipulators feels like it should have a root of `world` (or similar), not `base_link`.

For me, it comes down to whether the URDF describes a single "robot" that might be conceivably dropped into a larger workcell, or whether the URDF itself represents a higher-level workcell with multiple robots inside.

gavanderhoorn commented:
@JeremyZoss wrote:
> What does the phrase "shall not be changed" mean in the `flange` and `tool0` sections? 

it means that the transforms (ie: joints) defining those <del>may</del> should not be changed/updated by consumers of xacros/models.

> For the most part, all frames in the stock robot-model URDFs are considered to be static and un-changing.

Exactly, but I put that in there as I've seen people actually going in and changing the `rpy/xyz` values for `tool0` instead of defining new frames. It has the potential to break all sorts of things (especially generated code, such as IKFast plugins, see [Tool0 URDF Modifications Break Path Planning](https://answers.ros.org/question/256429) and https://github.com/ros-industrial/fanuc_experimental/issues/30 for example).

This way we can at least point to some documentation that makes this clear.

gavanderhoorn commented:
@JeremyZoss wrote:
> We don't currently have a `flange` model in most ROS-I URDFs, and in some cases may be using `tool0` for similar purposes (e.g. [abb\irb2400](https://github.com/ros-industrial/abb/blob/kinetic-devel/abb_irb2400_support/urdf/irb2400_12_155_macro.xacro#L150)).  

Yes, all our support pkgs (except `fanuc`, since ros-industrial/fanuc#213) are using `tool0` as what should really be `flange`. That is exactly the reason I <del>introduced</del> want to introduce the new frame: `tool0` is a mathematical concept, not an attachment point.

And I believe the awkwardness of having to undo the x+->z+ rotation shows that.

> How do we intend to handle this transition?  Some concerns:

If we are only talking about xacros we provide, I think there shouldn't be any bw-compatibility problems:

>   - changes to tool0 may affect existing models

The `tool0` defined in this REP is identical to the `tool0` that is already present in most (all?) of our robot support packages.

>   - how do we flag robots that have "compliant" and "non-compliant" tool0 definitions?

I'm not sure that is needed, as, as I wrote above, this REP does not propose any changes to the definition of `tool0`. But if desired, we could update the readmes of the respective support packages stating that they are "REP-XXXX compliant".

Consumers that are currently using `tool0` as an attachment point for their EEFs should not be affected by the introduction of this REP, or the addition of `flange` frames: `tool0` will stay defined as it is, we just update its semantics and intended usage. New composite xacros should be created by attaching EEF models to `flange`.

---

Edit; I think your questions warrant the addition of a "backwards compatibility" section (or similar) to the REP. I considered that, but I intend to submit this REP as a regular ROS REP, not a ROS-I REP, so I'm not sure such a section would make sense there. Unless we consider ROS-I as being part of ROS of course.

gavanderhoorn commented:
@JeremyZoss wrote:
> Do we need to recommend a standard naming convention for the TCP frame?  It's somewhat outside the scope of this document, since it won't be captured in the core robot URDF, but it's related to concerns about modularity and standardization.

hm, I had not considered that, but it could be interesting.

My first reaction would be that something like a TCP frame is typically part of the configuration of an application, and is thus something that is probably not hard-coded into the code that references/uses it. That leaves developers quite some freedom which we would be taking away if we start prescribing naming conventions for EEFs.

To me, naming conventions make the most sense for things that you cannot control or change, as everyone will have to live with the choices that others (essentially) made *for them*. If we agree on how we should all make those choices (or at least provide sufficient rationale) then not being able to change them should become less of an issue. That is what most REPs are about, so I think I would conclude that I don't feel we need to standardise TCP names in a REP.

gavanderhoorn commented:
@JeremyZoss wrote:
> We should include a recommendation on where the kinematic chain should __end__.  You recommend it use `base_link` as the root (okay), but do we recommend `flange` or `tool0` as the tip?  I recommend `flange`, as it seems the equivalent of `base_link`.  `base_link` and `flange` are both ROS-centric, geometry-based frames.

This is a good one. `flange` is indeed conceptually similar to `base_link`, but I think I would prefer `tool0` for the reason that industrial robots tend to 'plan' for that frame by default if no other is defined and / or selected (ie: all "world frame" Cartesian poses will show the pose of the `tool0` (or whatever it is called on a particular robot controller) if no other tool frame is currently defined or selected).

`flange` does not necessarily *have to* be coincident with `tool0` for all robot controllers. If we could guarantee that, then we could choose either.

> `base` and `tool0` seem more like virtual frames for association with robot controller kinematics/positioning.

that is basically true: using those frames will allow a correspondance between wat ROS consumes/produces and what the robot controller will use/show you.

I think I'd like to suggest we define `base_link->tool0` as the start and end of the default chain.

gavanderhoorn commented:
@JeremyZoss wrote:
> Do you need to add a discussion about intermediate frame/link names?  In general, there is more freedom/variation here, and most external ROS code/URDF should reference the base/tool frames instead.  However, it may still be good to standardize where possible.  We could recommend use of `link_1`...`link_6`, with exceptions to follow manufacturer-specific joint naming (e.g. motoman).

I purposefully did not include any naming for links other than the 'special' ones, as I felt it was out-of-scope. The REP is about 'coordinate frames', not 'link names' or naming in general.

Your motoman example however makes me reconsider this: if we do add something about naming intermediary links, I think I would like to suggest we prescribe a `link_N` scheme, and do not allow manufacturer-specific alternatives, for the very reason that they are *manufacturer-specific*. The point of this REP is to abstract away as much as possible any mfg specifics, and to increase reusability of components (and other xacros!) interfacing with robot models as much as possible by doing that. If we allow this exception, then we deliberately poke a hole in the abstraction, which is not something I think we should want.

We could argue that joint and link names are -- similar to TCP names -- part of the configuration of an application and should therefore not be strictly regulated. But I think that is actually not true, as it would fi make it harder (or it would at least not make it easier) to compose xacros if joint and link names are allowed to have arbitrary names.

gavanderhoorn commented:
@JeremyZoss wrote:
> In your Dual Manipulator chart, I disagree with the placement of `base_link`.  It feels like a true dual-arm manipulator (e.g. SDA20) might have a `base_link` as shown, along with other potential joints (e.g. `torso`).  However, a welding cell with two standalone manipulators feels like it should have a root of `world` (or similar), not `base_link`.
> 
> For me, it comes down to whether the URDF describes a single "robot" that might be conceivably dropped into a larger workcell, or whether the URDF itself represents a higher-level workcell with multiple robots inside.

I see your point, but we could say that no xacro is ever a 'top-level' xacro as long as it is a macro: if we define the higher-level workcell as a macro itself, I could instantiate it multiple times (perhaps in a 'factory' xacro) and then I would have multiple `world` frames, one for each cell.

And the `world` frame name is a bit arbitrary. I guess I didn't want to suddenly introduce any 'new' frame names in the examples to avoid confusion (hence the ellipses in the trees).

But as an example the contents of the section should be as clear as possible so your comments show that it needs to be improved. I'll try to do that.

gavanderhoorn commented:
@JeremyZoss wrote:

> There is potential confusion with references to `base_link` and "kinematic chain"
>   - You should clarify that we mean "___ROS___ kinematic chain" (no tie to robot kinematics)
>   - Remove the phrase "and forms the root of the kinematic chain..." from the initial `base_link` paragraph.  The key here is that this link is based on the physical robot geometry.
>   - Move the sentence "No special treatment ... root of kinematic chain..." from the `base_link` rationale section to a standalone paragraph above.  This is a recommendation for what to do with `base_link`, not a rationale for why we're doing that.

I think I agree with all your points, but wouldn't adding (as you suggest) "ROS" to "kinematic chain" wherever it is applicable fix most (all?) of the confusion? And make removing "and forms the root of the kinematic chain" from the initial `base_link` paragraph unnecessary?

gavanderhoorn commented:
Thanks for all your comments @JeremyZoss, you raised a lot of good points. I'll try to address your comments by updating the text.

Feel free to comment on my comments again. I hope I explained sufficiently why I wrote the REP as it is now.

gavanderhoorn committed:
Clarify base_link description and rationale.

gavanderhoorn commented:
@gavanderhoorn wrote:
> @JeremyZoss wrote:
> > We should include a recommendation on where the kinematic chain should __end__.  You recommend it use `base_link` as the root (okay), but do we recommend `flange` or `tool0` as the tip?  I recommend `flange`, as it seems the equivalent of `base_link`.  `base_link` and `flange` are both ROS-centric, geometry-based frames.
> 
> [..]
> 
> I think I'd like to suggest we define `base_link->tool0` as the start and end of the default chain.

Thinking about this some more: if I'd want to be consistent (and keep the ind robot controller as an example of sorts), the chain should actually be `base->tool0`.

Without some experimentation, I'm not sure I can say with certainty what would make more sense / work better though.

Levi-Armstrong commented on rep-xxxx.rst:
```diff
@@ -0,0 +1,268 @@
+
+
+flange
+''''''
```
Is the addition of the flang link just so there is link with the ros convention at the end of the kinematic change?

gavanderhoorn replied:
Partly, yes.

But the rationale for having a link that is compliant with REP-103 is that it makes attaching EEF subhierarchies very convenient (ROS 'attaches' subtrees to x+ by default). `tool0` has z+ forward, which leads to your EEF model ending up rotated in weird / unexpected ways wrt your flange.

To fix that you'd have to undo the rotation that `tool0` introduces, which is what I've observed ppl do. By introducing `flange` we avoid all that.

gavanderhoorn replied:
And it resolves the strange mixup of a physical attachment point (`flange`) and the mathematical concept that is the tool centre point (`tool0`) by separating them into two distinct frames.

Levi-Armstrong commented:
:+1:

gavanderhoorn committed:
wip

gavanderhoorn committed:
This REP is only for serial industrial robots.

Parallel configurations are not considered.

gavanderhoorn committed:
Minor formatting.

gavanderhoorn committed:
Add more examples, improve those that we already had.

gavanderhoorn committed:
Add a Definitions section.

gavanderhoorn committed:
Add note about robots with left-handed coord systems.

gavanderhoorn committed:
further wip.

gavanderhoorn committed:
Update date.

gavanderhoorn committed:
Questions resolved.

 - q1: tool0 will be child of flange
 - q2: we will discourage using 'toolN' names for frames

gavanderhoorn committed:
Complete description of link frames.

gavanderhoorn commented:
I'm going to close this, as there have quite a few commits to this branch since the last discussion took place.

I'll open a new final discussion PR when I'm almost ready to submit this REP to `ros-infrastructure/rep` for the public review.

gavanderhoorn commented:
Thanks to all who provided input. That was (and is) much appreciated.
